### PR TITLE
Fixes #310, adds softmax,logsoftmax,logistic and bce losses

### LINKED ISSUE #310: no logp support for 3D KnetArrays
Current CUDNN implementation of `logp` does not support 3D KnetArrays.

The results are not correct for ` logp(x::KnetArray{Float64,3},1)`.
In addition, there is a bug in ` logp(x::KnetArray{Float64,3},2)`. 

```JULIA
julia> x = randn(2,2,2)
2×2×2 Array{Float64,3}:
[:, :, 1] =
 0.0196004  0.0178769
 0.548913   1.26022

[:, :, 2] =
  0.449092   0.81334
 -1.95281   -1.43875

julia> logp(x,1) - Array(logp(KnetArray(x),1))
2×2×2 Array{Float64,3}:
[:, :, 1] =
 0.975221  0.473384
 0.975221  0.473384

[:, :, 2] =
 0.899652  0.522074
 0.899652  0.522074

julia> logp(x,2) - Array(logp(KnetArray(x),2))
ERROR: Transpose is supported only for 2D KnetArrays
Stacktrace:
 [1] transpose(::Knet.KnetArray{Float64,3}) at /scratch/users/eakyurek13/.julia/v0.6/Knet/src/linalg.jl:89
 [2] logp(::Knet.KnetArray{Float64,3}, ::Int64, ::Vararg{Int64,N} where N) at /scratch/users/eakyurek13/.julia/v0.6/Knet/src/loss.jl:19
```

### PR DESCRIPTION
**Pull request by ekinakyurek**
Fixes #310, adds softmax,logsoftmax,logistic loss and binary-cross-entropy functionalities
Co-authored-by: CarloLucibello <carlolucibello@users.noreply.github.com>

@CarloLucibello I created a clean and single commit branch from our work. I used co-authorship functionality of GitHub. Could you okey with this commit? 

The only difference is repeating dims: `dims=(1,1)` and single dims in an array/tuple: `dims=(1)`, but I think it is not necessary to handle(to send gpu) these odd cases.

### CODE REVIEW

**Comment by denizyuret** [comment]
@CarloLucibello I think this PR has all the functionality and attribution to you and @ekinakyurek -- can I just merge this?

**Comment by CarloLucibello** [comment]
There is some little feature missing, like the possibility of equivalently using the arguments `dims=1`, `dims=(1,)` and `dims=[1]` to obtain the same results (this is also the `sum` behavior). 
Most importantly, I would like all the tests I added in #313  to be added here if this gets merged.
I think it is easier to merge #313, it is ready and I tested it both on cpu and gpu,
but I'm fine with either options as long as the tests get in.

**Comment by ekinakyurek** [comment]
I didn't see the tests at #313, I recovered them now. I also recovered `dimvec(x, dims)` from #313 clearing some unnecessary size calls.

For the clarity: The previous commit also handles `dims` in the same way with `sum`. However, it doesn't send these strange cases to cudnn calls e.g. `dims=(1,)`, `dims=(1,1)`, ` dims=(1,2,3)` where `ndims(x)=3`. Since the speed matters, I recovered them in this commit.
